{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyP4KNi801iHDGpGsaPo80p6"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["[예외객체]\n","- Exception을 사용하면 예외의 종류에 대해 알 수 있다."],"metadata":{"id":"u1UYBr8XNE9C"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"fXmEoFidNEGY"},"outputs":[],"source":["try:\n","    x, y = map(int, input('밑변과 높이를 입력해주세요').split(\" \"))\n","    print('삼각형의 넓이는 {0}입니다'.format(x*y/2))\n","except:\n","    print(\"정수로 입력하지 않아 계산이 불가합니다.\")"]},{"cell_type":"code","source":["try:\n","    x, y = map(int, input('밑변과 높이를 입력해주세요').split(\" \"))\n","    print('삼각형의 넓이는 {0}입니다'.format(x*y/2))\n","except Exception as exception:\n","    print(\"type(exception):\", type(exception))\n","    print(\"exception:\", exception)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"12W9NCH8Nivv","executionInfo":{"status":"ok","timestamp":1723594181989,"user_tz":-540,"elapsed":47640,"user":{"displayName":"Inceha","userId":"13455914109043978432"}},"outputId":"582095a3-57f0-4af9-ec50-b17798a9d8db"},"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["밑변과 높이를 입력해주세요sdf\n","type(exception): <class 'ValueError'>\n","exception: invalid literal for int() with base 10: 'sdf'\n"]}]},{"cell_type":"markdown","source":["예외객체 자료형은 Value Error이고  \n","내용은 invalid literal for int() with base 10: 'sdf'이다.  \n","\n","[예외 구분하기]\n","여러 가지 예외가 발생하는 상황을 구분지어 대비하기 위함"],"metadata":{"id":"vziFm_MYOEFB"}},{"cell_type":"code","source":["numbers = [23, 11, 7, 4, 12]\n","\n","try:\n","    number_input = int(input(\"찾고싶은 값의 위치를 입력해주세요.\"))\n","    print(f'{number_input}번째 요소 : {numbers[number_input]}')\n","except Exception as exception:\n","    print(\"type(exception):\", type(exception))\n","    print(\"exception:\", exception)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SQgh_GnKOStt","executionInfo":{"status":"ok","timestamp":1723594413861,"user_tz":-540,"elapsed":2389,"user":{"displayName":"Inceha","userId":"13455914109043978432"}},"outputId":"8b8233b5-51cc-483e-fb2d-fcda9de4bbe5"},"execution_count":4,"outputs":[{"output_type":"stream","name":"stdout","text":["찾고싶은 값의 위치를 입력해주세요.2\n","2번째 요소 : 7\n"]}]},{"cell_type":"markdown","source":["[except : list index out of range]와 같은 에러를 구분하고처리할 수 있는 기능을 파이썬에서 제공한다.  \n","\n","```\n","try:\n","    예외가 발생할 가능성이 있는 구문\n","except 예외 a:\n","    예외 a가 발생했을 때 실행할 구문\n","except 예외 b:\n","    예외 b가 발생했을 때 실행할 구문\n","except 예외 c:\n","    예외 c가 발생했을 때 실행할 구문\n","```"],"metadata":{"id":"X-Fe6YUaO7gd"}},{"cell_type":"code","source":["numbers = [23, 11, 7, 4, 12]\n","\n","try:\n","    number_input = int(input(\"찾고싶은 값의 위치를 입력해주세요.\"))\n","    print(f'{number_input}번째 요소 : {numbers[number_input]}')\n","except ValueError:\n","    print(\"정수로 입력해주세요\")\n","except IndexError:\n","    print(\"리스트의 범위를 벗어났습니다. 입력값을 다시 확인해주세요.\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0vwaJnKOPXG8","executionInfo":{"status":"ok","timestamp":1723594658009,"user_tz":-540,"elapsed":3588,"user":{"displayName":"Inceha","userId":"13455914109043978432"}},"outputId":"a912dae0-9195-417c-fd94-9384de94ade1"},"execution_count":6,"outputs":[{"output_type":"stream","name":"stdout","text":["찾고싶은 값의 위치를 입력해주세요.300\n","리스트의 범위를 벗어났습니다. 입력값을 다시 확인해주세요.\n"]}]},{"cell_type":"markdown","source":["[예외 구분 구문과 예외 객체]\n","as 키워드를 이용해 as exception을 사용할 수 있음"],"metadata":{"id":"78VqipVNP7fV"}},{"cell_type":"code","source":["numbers = [23, 11, 7, 4, 12]\n","\n","try:\n","    number_input = int(input(\"찾고싶은 값의 위치를 입력해주세요.\"))\n","    print(f'{number_input}번째 요소 : {numbers[number_input]}')\n","except ValueError as exception:\n","    print(\"정수로 입력해주세요\")\n","    print(\"Exception:\", exception)\n","except IndexError as exception:\n","    print(\"리스트의 범위를 벗어났습니다. 입력값을 다시 확인해주세요.\")\n","    print(\"Exception:\", exception)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SqrC9FsQQA90","executionInfo":{"status":"ok","timestamp":1723594768440,"user_tz":-540,"elapsed":2461,"user":{"displayName":"Inceha","userId":"13455914109043978432"}},"outputId":"1fb94b5d-24a3-41a7-8ccf-ae04ecd48cf7"},"execution_count":7,"outputs":[{"output_type":"stream","name":"stdout","text":["찾고싶은 값의 위치를 입력해주세요.1000\n","리스트의 범위를 벗어났습니다. 입력값을 다시 확인해주세요.\n","Exception: list index out of range\n"]}]},{"cell_type":"markdown","source":["프로그램 실행 중 위 2가지 예외처리로 처리가 불가능한 경우를 대비해 앞서 사용했던  \n","모든 예외처리가 가능한 Exception을 이용해 대처할 수 있다."],"metadata":{"id":"XtEcjg_VQRBp"}},{"cell_type":"code","source":["numbers = [23, 11, 7, 4, 12]\n","\n","try:\n","    number_input = int(input(\"찾고싶은 값의 위치를 입력해주세요.\"))\n","    print(f'{number_input}번째 요소 : {numbers[number_input]}')\n","    에러를.만들기위한\n","except ValueError as exception:\n","    print(\"정수로 입력해주세요\")\n","    print(\"Exception:\", exception)\n","except IndexError as exception:\n","    print(\"리스트의 범위를 벗어났습니다. 입력값을 다시 확인해주세요.\")\n","    print(\"Exception:\", exception)\n","except Exception as exception:\n","    print(\"대처 불가한 예외가 발생했습니다.\")\n","    print(\"Exception:\", exception)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"cz1llfnmQZX_","executionInfo":{"status":"ok","timestamp":1723594923735,"user_tz":-540,"elapsed":2503,"user":{"displayName":"Inceha","userId":"13455914109043978432"}},"outputId":"0b4a00e3-3975-48de-eb0b-80d76d4be1f1"},"execution_count":11,"outputs":[{"output_type":"stream","name":"stdout","text":["찾고싶은 값의 위치를 입력해주세요.3\n","3번째 요소 : 4\n","대체 불가한 예외가 발생했습니다.\n","Exception: name '에러를' is not defined\n"]}]},{"cell_type":"markdown","source":["####예외 구분 구문의 이해\n","- 예외 종류와 그에 대한 설명(구글링 필요)"],"metadata":{"id":"LvxeO-LRRLiG"}},{"cell_type":"code","source":["#1. 예외 종류와 그에 대한 설명\n","# KeyError : 주로 딕셔너리에서 발생하는 에러로, 없는 Key 값에 접근하려 할 때 발생\n","# ValueError : 부적절한 값을 가진 인자를 받았을 때 발생하는 에러\n","# SyntaxError : 문법 오류, 조건문에서 :을 빼먹거나 문법에 오류가 있을 때 발생\n","# NameError : 지역변수, 전역변수의 이름을 찾을 수 없는 경우 발생\n","# ZeroDivisionError : 숫자를 0으로 나누려할 때 발생하는 에러\n","# FileNotFoundError : 존재하지 않는 파일이나 디렉토리에 접근하려 할 때\n","# TypeError : 서로 다른 타입으로 연산하려고 할 때\n","# AttributeError : 잘못된 메서드나 속성을 호출하거나 대입했을 때 발생\n","# ConnectionError : 서버를 켜지 않았을 때 발생"],"metadata":{"id":"NQx7miLeRRat"},"execution_count":null,"outputs":[]}]}